### PR TITLE
Multi-GPU support in GPUPredictor.

### LINKED ISSUE #3756: Multi-gpu terminate called after throwing an instance of 'thrust::system::system_error'  parallel_for failed: out of memory
Building xgboost for multi-gpu support  from https://github.com/dmlc/xgboost/commit/c6b5df67f64e9e5feba29547faea03dd0ce707cc
With CUDA 9.0
With NCCL 2.2.13-1+cuda9.0
on amazon p2.8xlarge 488GB Ram , 32 vcpus, 8 gpus K80

CODE

    print(X.count())
    print(X.info(memory_usage='deep'))

    data = Data(X, y)
    dtrain = xgb.DMatrix(data.X_train, data.y_train)
    dval = xgb.DMatrix(data.X_val, data.y_val)
    p = {'min_child_weight': 50,
        'subsample': 1.0,
        'max_depth': 11,
        'learning_rate': 0.03,
        'colsample_bytree': 0.4,
        'sketch_eps': 0.01,
        'n_gpus': 8,
        'alpha': 0,
        'lambda': 1,
        'max_bin': 64,
        'tree_method': 'gpu_hist',
        'objective': 'gpu:reg:logistic'}
    bst = xgb.train(p, dtrain, 10, [(dtrain, "train"), (dval, "val")])

DATAFRAME has 120M entries 108 features (doubles)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 124646054 entries, 0 to 139271
Columns: 108 entries, xxxxx to yyyyyyy
dtypes: float32(108)
memory usage: 51.1 GB

Traceback (most recent call last):
  File "train_now.py", line 120, in <module>
    main()
  File "train_now.py", line 87, in main
    bst = xgb.train(params, dtrain, args.num_rounds, [(dtrain, "train"), (dval, "val")])
  File "/mnt/xgboost/venv/local/lib/python2.7/site-packages/xgboost/training.py", line 216, in train
    xgb_model=xgb_model, callbacks=callbacks)
  File "/mnt/xgboost/venv/local/lib/python2.7/site-packages/xgboost/training.py", line 74, in _train_internal
    bst.update(dtrain, i, obj)
  File "/mnt/xgboost/venv/local/lib/python2.7/site-packages/xgboost/core.py", line 1035, in update
    dtrain.handle))
  File "/mnt/xgboost/venv/local/lib/python2.7/site-packages/xgboost/core.py", line 165, in _check_call
    raise XGBoostError(_LIB.XGBGetLastError())
xgboost.core.XGBoostError: [19:26:13] /mnt/dmlc/xgboost/include/xgboost/../../src/common/common.h:41: /mnt/dmlc/xgboost/src/predictor/../common/device_helpers.cuh: 409: out of memory

Stack trace returned 10 entries:
[bt] (0) /mnt/xgboost/venv/xgboost/libxgboost.so(dmlc::StackTrace[abi:cxx11]()+0x5b) [0x7fe98b5dba6b]
[bt] (1) /mnt/xgboost/venv/xgboost/libxgboost.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x28) [0x7fe98b5dc2e8]
[bt] (2) /mnt/xgboost/venv/xgboost/libxgboost.so(dh::ThrowOnCudaError(cudaError, char const*, int)+0x22f) [0x7fe98b814e4f]
[bt] (3) /mnt/xgboost/venv/xgboost/libxgboost.so(xgboost::predictor::DeviceMatrix::DeviceMatrix(xgboost::DMatrix*, int, bool)+0x16f) [0x7fe98b84ed4f]
[bt] (4) /mnt/xgboost/venv/xgboost/libxgboost.so(xgboost::predictor::GPUPredictor::DevicePredictInternal(xgboost::DMatrix*, xgboost::HostDeviceVector<float>*, xgboost::gbm::GBTreeModel const&, unsigned long, unsigned long)+0xbc7) [0x7fe98b850d47]
[bt] (5) /mnt/xgboost/venv/xgboost/libxgboost.so(xgboost::predictor::GPUPredictor::UpdatePredictionCache(xgboost::gbm::GBTreeModel const&, std::vector<std::unique_ptr<xgboost::TreeUpdater, std::default_delete<xgboost::TreeUpdater> >, std::allocator<std::unique_ptr<xgboost::TreeUpdater, std::default_delete<xgboost::TreeUpdater> > > >*, int)+0x75) [0x7fe98b851475]
[bt] (6) /mnt/xgboost/venv/xgboost/libxgboost.so(xgboost::gbm::GBTree::DoBoost(xgboost::DMatrix*, xgboost::HostDeviceVector<xgboost::detail::GradientPairInternal<float> >*, xgboost::ObjFunction*)+0x6d5) [0x7fe98b661ef5]
[bt] (7) /mnt/xgboost/venv/xgboost/libxgboost.so(xgboost::LearnerImpl::UpdateOneIter(int, xgboost::DMatrix*)+0x361) [0x7fe98b66f351]
[bt] (8) /mnt/xgboost/venv/xgboost/libxgboost.so(XGBoosterUpdateOneIter+0x48) [0x7fe98b5cfb58]
[bt] (9) /usr/lib/x86_64-linux-gnu/libffi.so.6(ffi_call_unix64+0x4c) [0x7fe9fd287e40]

terminate called after throwing an instance of 'thrust::system::system_error'
  what():  parallel_for failed: out of memory
Aborted

nvidia-smi snapshot before the crash

    +-----------------------------------------------------------------------------+
    | NVIDIA-SMI 390.87                 Driver Version: 390.87                    |
    |-------------------------------+----------------------+----------------------+
    | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
    | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
    |===============================+======================+======================|
    |   0  Tesla K80           Off  | 00000000:00:17.0 Off |                    0 |
    | N/A   57C    P0    69W / 149W |   2095MiB / 11441MiB |      0%      Default |
    +-------------------------------+----------------------+----------------------+
    |   1  Tesla K80           Off  | 00000000:00:18.0 Off |                    0 |
    | N/A   53C    P0    82W / 149W |   2095MiB / 11441MiB |      0%      Default |
    +-------------------------------+----------------------+----------------------+
    |   2  Tesla K80           Off  | 00000000:00:19.0 Off |                    0 |
    | N/A   57C    P0    65W / 149W |   2095MiB / 11441MiB |      0%      Default |
    +-------------------------------+----------------------+----------------------+
    |   3  Tesla K80           Off  | 00000000:00:1A.0 Off |                    0 |
    | N/A   53C    P0    81W / 149W |   2095MiB / 11441MiB |      0%      Default |
    +-------------------------------+----------------------+----------------------+
    |   4  Tesla K80           Off  | 00000000:00:1B.0 Off |                    0 |
    | N/A   59C    P0    70W / 149W |   2095MiB / 11441MiB |      0%      Default |
    +-------------------------------+----------------------+----------------------+
    |   5  Tesla K80           Off  | 00000000:00:1C.0 Off |                    0 |
    | N/A   53C    P0    82W / 149W |   2095MiB / 11441MiB |      0%      Default |
    +-------------------------------+----------------------+----------------------+
    |   6  Tesla K80           Off  | 00000000:00:1D.0 Off |                    0 |
    | N/A   59C    P0    73W / 149W |   2095MiB / 11441MiB |      0%      Default |
    +-------------------------------+----------------------+----------------------+
    |   7  Tesla K80           Off  | 00000000:00:1E.0 Off |                    0 |
    | N/A   54C    P0    87W / 149W |   2095MiB / 11441MiB |      0%      Default |
    +-------------------------------+----------------------+----------------------+

    +-----------------------------------------------------------------------------+
    | Processes:                                                       GPU Memory |
    |  GPU       PID   Type   Process name                             Usage      |
    |=============================================================================|
    |    0     35738      C   -                                           2082MiB |
    |    1     35738      C   -                                           2082MiB |
    |    2     35738      C   -                                           2082MiB |
    |    3     35738      C   -                                           2082MiB |
    |    4     35738      C   -                                           2082MiB |
    |    5     35738      C   -                                           2082MiB |
    |    6     35738      C   -                                           2082MiB |
    |    7     35738      C   -                                           2082MiB |
    +-----------------------------------------------------------------------------+

I can not publish the dataset, but creating a similar one should be no problem.

Thanks

### PR DESCRIPTION
**Pull request by canonizer**
- GPUPredictor is multi-GPU
- removed DeviceMatrix, as it has been made obsolete by using HostDeviceVector in DMatrix

Closes #3756

### CODE REVIEW

File: src/predictor/gpu_predictor.cu [hunk 1]
```diff
@@ -262,31 +222,81 @@ class GPUPredictor : public xgboost::Predictor {
...
-                             HostDeviceVector<bst_float>* out_preds,
-                             const gbm::GBTreeModel& model, size_t tree_begin,
-                             size_t tree_end) {
-    if (tree_end - tree_begin == 0) {
-      return;
+  void DeviceOffsets(const HostDeviceVector<size_t>& data, std::vector<size_t>* out_offsets) {
+    auto& offsets = *out_offsets;
+    offsets.resize(devices.Size() + 1);
+    offsets[0] = 0;
+#pragma omp parallel for schedule(static, 1) if (devices.Size() > 1)
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu]: Sorry for dropping by.  Just to be safe, it might be better to save the current device before spawning threads that can change it.  Otherwise subsequent code could potentially access memory at wrong device.

https://github.com/dmlc/xgboost/blob/baef5741dfeec121dbeca9802e6b00a75ed72d03/src/common/device_helpers.cuh#L947

Comment (canonizer) [reply src/predictor/gpu_predictor.cu]: `cudaSetDevice()` changes the device of the caller thread only, so it does not alter the device in a different thread.

In case only a single device is used (and no threads are spawned), `cudaSetDevice()` will be called with that device.

Otherwise, the code using a GPU is responsible for setting the device being used. This means `cudaSetDevice()` in public methods, in shards and in OpenMP loops (with 1 iteration per device). Private methods can assume that the right device has been set already.

**Comment by hcho3** [comment]
@canonizer Can you add a test for multi-GPU prediction? I am about to add a multi-GPU slave worker to the Jenkins CI server. The multi-GPU tests will run as a separate task than single-GPU tests.

**Comment by hcho3** [comment]
Any updates on multi-GPU tests?

File: src/predictor/gpu_predictor.cu [hunk 2]
```diff
@@ -143,19 +100,21 @@ struct DevicePredictionNode {
 
 struct ElementLoader {
   bool use_shared;
-  size_t* d_row_ptr;
-  Entry* d_data;
+  const size_t* d_row_ptr;
```
[the remark this answers is not in the document]
Comment (canonizer) [reply src/predictor/gpu_predictor.cu]: Done.

File: src/predictor/gpu_predictor.cu [hunk 3]
```diff
@@ -225,14 +184,15 @@ __device__ float GetLeafWeight(bst_uint ridx, const DevicePredictionNode* tree,
 template <int BLOCK_THREADS>
 __global__ void PredictKernel(const DevicePredictionNode* d_nodes,
                               float* d_out_predictions, size_t* d_tree_segments,
-                              int* d_tree_group, size_t* d_row_ptr,
-                              Entry* d_data, size_t tree_begin,
+                              int* d_tree_group, const size_t* d_row_ptr,
```
[the remark this answers is not in the document]
Comment (canonizer) [reply src/predictor/gpu_predictor.cu]: Done.

File: src/predictor/gpu_predictor.cu [hunk 4]
```diff
@@ -178,9 +137,9 @@ struct ElementLoader {
       return smem[threadIdx.x * num_features + fidx];
     } else {
       // Binary search
-      auto begin_ptr = d_data + d_row_ptr[ridx];
-      auto end_ptr = d_data + d_row_ptr[ridx + 1];
-      Entry* previous_middle = nullptr;
+      auto begin_ptr = d_data + d_row_ptr[ridx] - entry_start;
```
[the remark this answers is not in the document]
Comment (canonizer) [reply src/predictor/gpu_predictor.cu]: Yes, but probably in a different pull request.

If you want it in this pull request, feel free to comment on this, and I'll get back to it on Thursday.

**Comment by canonizer** [comment]
Added a multi-GPU test for GPUPredictor and addressed reviewers' comments.

File: src/predictor/gpu_predictor.cu (hunk 4 above)
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/predictor/gpu_predictor.cu]: No need to rush into cleaning up.  We can do it later. :)

File: src/predictor/gpu_predictor.cu [hunk 5]
```diff
@@ -175,9 +137,9 @@ struct ElementLoader {
       return smem[threadIdx.x * num_features + fidx];
     } else {
       // Binary search
-      auto begin_ptr = d_data + d_row_ptr[ridx];
-      auto end_ptr = d_data + d_row_ptr[ridx + 1];
-      Entry* previous_middle = nullptr;
+      auto begin_ptr = d_data.begin() + d_row_ptr[ridx] - entry_start;
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/predictor/gpu_predictor.cu]: I think you can make this change yourself now that you are a member.

**Comment by trivialfis** [comment]
Okay, I fixed a bug that's caught by my profiling script.  But I can't reproduce the one on Jenkins.

**Comment by trivialfis** [comment]
@hcho3 Thanks!  I ran cuda-memcheck and Sanitizer with no luck so far.  The test requires very small amount of memory.

**Comment by hcho3** [comment]
I compiled and ran this pull request on my p2.8xlarge instance and got the same error. I will run it through cuda-gdb to see if it helps.

**Comment by hcho3** [comment]
@trivialfis @canonizer I got this backtrace by running `testxgboost` through gdb:
```
#0  __memmove_avx_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-avx-unaligned.S:136
#1  0x00007ffff5958bdf in ?? () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#2  0x00007ffff5a123ee in ?? () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#3  0x00007ffff5a126ed in ?? () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#4  0x00007ffff5a139b6 in ?? () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#5  0x00007ffff59256be in ?? () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#6  0x00007ffff59259d8 in ?? () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#7  0x00007ffff5a761d5 in cuMemcpy () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#8  0x000000000088f022 in cudart::driverHelper::memcpyDispatch(void*, void const*, unsigned long, cudaMemcpyKind, bool) ()
#9  0x000000000086f896 in cudart::cudaApiMemcpy(void*, void const*, unsigned long, cudaMemcpyKind) ()
#10 0x0000000000891ed8 in cudaMemcpy ()
#11 0x0000000000779b52 in xgboost::predictor::GPUPredictor::DeviceOffsets () at /home/ubuntu/xgboost/src/predictor/gpu_predictor.cu:234
#12 0x00007ffff712d43e in ?? () from /usr/lib/x86_64-linux-gnu/libgomp.so.1
#13 0x00007ffff6cf26ba in start_thread (arg=0x7ffff25d7700) at pthread_create.c:333
#14 0x00007ffff6a2841d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

Maybe this line is problematic?
https://github.com/dmlc/xgboost/blob/10bc710098f75298c3c3ca55b568c623cd5a508f/src/predictor/gpu_predictor.cu#L233-L236

**Comment by hcho3** [comment]
I added some diagnostic printout before the cudaMemcpy line:
```
[08:44:06] /home/ubuntu/xgboost/src/predictor/gpu_predictor.cu:233: cudaMemcpy(&offsets[0 + 1], 0x7c136c0200 + 2 - 1, sizeof(size_t), cudaMemcpyDefault));
[08:44:06] /home/ubuntu/xgboost/src/predictor/gpu_predictor.cu:233: cudaMemcpy(&offsets[1 + 1], 0x7c13ec0200 + 2 - 1, sizeof(size_t), cudaMemcpyDefault));
[08:44:06] /home/ubuntu/xgboost/src/predictor/gpu_predictor.cu:233: cudaMemcpy(&offsets[2 + 1], 0x7c13ac0200 + 2 - 1, sizeof(size_t), cudaMemcpyDefault));
[08:44:06] /home/ubuntu/xgboost/src/predictor/gpu_predictor.cu:233: cudaMemcpy(&offsets[3 + 1], 0x7c132c0200 + 2 - 1, sizeof(size_t), cudaMemcpyDefault));
[08:44:06] /home/ubuntu/xgboost/src/predictor/gpu_predictor.cu:233: cudaMemcpy(&offsets[4 + 1], 0x7c142c0200 + 2 - 1, sizeof(size_t), cudaMemcpyDefault));
[08:44:06] /home/ubuntu/xgboost/src/predictor/gpu_predictor.cu:233: cudaMemcpy(&offsets[5 + 1], 0x7c18380000 + 1 - 1, sizeof(size_t), cudaMemcpyDefault));
[08:44:07] /home/ubuntu/xgboost/src/predictor/gpu_predictor.cu:233: cudaMemcpy(&offsets[6 + 1], 0 + 0 - 1, sizeof(size_t), cudaMemcpyDefault));
```
The last cudaMemcpy call fails because the device pointer for shard 6 is null.

File: tests/cpp/predictor/test_gpu_predictor.cu [hunk 6]
```diff
@@ -76,5 +76,46 @@ TEST(gpu_predictor, Test) {
...
+  std::vector<std::unique_ptr<RegTree>> trees;
+  trees.push_back(std::unique_ptr<RegTree>(new RegTree()));
+  trees.back()->InitModel();
+  (*trees.back())[0].SetLeaf(1.5f);
+  (*trees.back()).Stat(0).sum_hess = 1.0f;
+  gbm::GBTreeModel model(0.5);
+  model.CommitModel(std::move(trees), 0);
+  model.param.num_output_group = 1;
+
+  int n_row = 5;
```
Comment (hcho3) [inline tests/cpp/predictor/test_gpu_predictor.cu]: Changing this line to
```suggestion
  int n_row = 8;
```
gets rid of segmentation fault.

The p2.8xlarge instance has 8 GPUs, so with the matrix with 5 rows, some of the GPUs were getting 0 row. We should handle this edge case either by restricting the number of devices when too few rows are given, or by correctly handling zero-row shards.

Comment (hcho3) [reply tests/cpp/predictor/test_gpu_predictor.cu]: @trivialfis How should we handle this edge case?

Comment (trivialfis) [reply tests/cpp/predictor/test_gpu_predictor.cu]: @hcho3 cuda-gdb doesn't work for me with NCCL, neither on Fedora nor on Ubuntu.. :(
The `GPUSet::All()` has an optional parameter specifying number of rows.

**Comment by trivialfis** [comment]
@hcho3 This again comes down to changing parameters.  We need to handle the situation when `n_gpus` is limited hence changed by the `n_rows` of input data.

**Comment by trivialfis** [comment]
@canonizer , @hcho3 , @RAMitchell  I tried to overcome it with a check to see if `GPUSet` is changed, if so all `DeviceShard`s are re-built.  It might not be a nice solution, suggestions are welcomed.

File: tests/cpp/predictor/test_gpu_predictor.cu (hunk 6 above)
Comment (trivialfis) [reply tests/cpp/predictor/test_gpu_predictor.cu]: Done.

**Comment by trivialfis** [comment]
@hcho3 Hi, could you give another look before I merge it?

File: src/objective/multiclass_obj.cu [hunk 7]
```diff
@@ -65,8 +65,8 @@ class SoftmaxMultiClassObj : public ObjFunction {
     const auto ndata = static_cast<int64_t>(preds.Size() / nclass);
 
     // clear out device memory;
-    out_gpair->Reshard(GPUSet::Empty());
-    preds.Reshard(GPUSet::Empty());
+    // out_gpair->Reshard(GPUSet::Empty());
+    // preds.Reshard(GPUSet::Empty());
```
Comment (hcho3) [inline src/objective/multiclass_obj.cu]: Why are we commenting out lines? If they are not needed, we should just remove them

Comment (trivialfis) [reply src/objective/multiclass_obj.cu]: Done.

File: tests/cpp/predictor/test_gpu_predictor.cu [hunk 8]
```diff
@@ -9,12 +9,14 @@
...
 namespace xgboost {
 namespace predictor {
+
 TEST(gpu_predictor, Test) {
   std::unique_ptr<Predictor> gpu_predictor =
       std::unique_ptr<Predictor>(Predictor::Create("gpu_predictor"));
   std::unique_ptr<Predictor> cpu_predictor =
       std::unique_ptr<Predictor>(Predictor::Create("cpu_predictor"));
 
+  // gpu_predictor->Init({std::pair<std::string, std::string>("n_gpus", "1")}, {});
```
Comment (hcho3) [inline tests/cpp/predictor/test_gpu_predictor.cu]: Remove this line too

Comment (trivialfis) [reply tests/cpp/predictor/test_gpu_predictor.cu]: Thank for cleaning up for me.

**Comment by hcho3** [comment]
Thanks everyone!
